src/realm.h

```diff
@@ -1798,4 +1791,255 @@ RLM_API realm_notification_token_t* realm_results_add_notification_callback(real
+    RLM_HTTP_PUT,
+    RLM_HTTP_DELETE,
+} realm_http_method_e;
+
+typedef struct realm_http_header {
+    const char* name;
+    const char* value;
+} realm_http_header_t;
+
+typedef struct realm_http_request {
+    realm_http_method_e method;
+    const char* url;
+    uint64_t timeout_ms;
+    const realm_http_header_t* headers;
+    size_t num_headers;
+    const char* body;
+    size_t body_size;
+    bool uses_refresh_token;
+} realm_http_request_t;
+
+typedef struct realm_http_response {
+    int status_code;
+    int custom_status_code;
+    const realm_http_header_t* headers;
+    size_t num_headers;
+    const char* body;
+    size_t body_size;
+} realm_http_response_t;
+
+typedef void (*realm_http_completion_func_t)(void* userdata, const realm_http_response_t*);
+typedef void (*realm_http_request_func_t)(void* transport_userdata, const realm_http_request_t*, void* userdata,
+                                          realm_http_completion_func_t response_received,
+                                          realm_free_userdata_func_t free_callback_data);
+
+typedef bool (*realm_websocket_on_open_func_t)(void* userdata);
+typedef bool (*realm_websocket_on_close_func_t)(void* userdata, bool was_clean, uint8_t code, const char* reason);
+typedef bool (*realm_websocket_on_error_func_t)(void* userdata, void* socket_userdata);
+typedef bool (*realm_websocket_on_message_func_t)(void* userdata, const uint8_t* data, size_t num_bytes,
+                                                  bool is_text);
+
+/**
+ * Create a new websocket.
+ *
+ * The function is expected to return an opaque handle representing a native
+ * WebSocket implementation, which also takes ownership of @a userdata and must
+ * take care to call @a free_func when the websocket handle is released.
+ *
+ * @param transport_userdata The userdata associated with a
+ *                           `realm_async_io_interface_t` object.
+ * @param url Zero-terminated URL string.
+ * @param protocols Comma-separated zero-terminated string of protocol names.
+ * @param userdata Caller-provided handle to be associated with this websocket
+ *                 and passed to event callbacks, such as
+ *                 `realm_websocket_on_open_func_t` etc.
+ * @param free_func Caller-provided function pointer that should be invoked when
+ *                  the websocket is destroyed, with @a userdata as the
+ *                  argument.
+ * @returns An opaque handle that will be passed to other callbacks, such as
+ *          `realm_websocket_send_binary_func_t` etc., or NULL if an error
+ *          occurred. If an error happened, API calls relying on this function
+ *          will report `RLM_ERR_CALLBACK`, and it is up to the implementation
+ *          to surface detailed information about the error (e.g. through the
+ *          network transport userdata pointer).
+ */
+typedef void* (*realm_websocket_new_func_t)(void* io_interface_userdata, const char* url, const char* protocols,
```

Comment:
Interesting, is this pre-empting abstracting the sync networking layer? 😈

Reply:
Yes. :) I took the opportunity to think a bit about that as well.

The API is very inspired by [the Emscripten interface](https://github.com/emscripten-core/emscripten/blob/master/system/include/emscripten/websocket.h).